src/core/varlink-metrics.c

```diff
@@ -1,11 +1,30 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
+#include "hashmap.h"
 #include "json-util.h"
 #include "manager.h"
+#include "list.h"
+#include "set.h"
 #include "sd-json.h"
 #include "sd-varlink.h"
+#include "unit-def.h"
 #include "varlink-metrics.h"
 
+static unsigned get_list_size(Manager *m, UnitType t);
+
+static unsigned get_list_size(Manager *m, UnitType t) {
+        unsigned count = 0;
+
+        assert(m);
+
+        // TODO: check t >= _UNIT_TYPE_MAX
+
+        LIST_FOREACH(units_by_type, _u, m->units_by_type[t])
```

Comment:
here you do N iterations (once per unit type). I think you should do one pass and just count different unit types.

You can use
```
size_t units[_UNIT_TYPE_MAX];
```